cpp/src/arrow/dataset/file_base.cc

```diff
@@ -222,9 +214,8 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
 }
 
 Result<std::shared_ptr<FileSystemDataset>> FileSystemDataset::Write(
-    const WritePlan& plan, std::shared_ptr<ScanContext> scan_context) {
-  std::vector<std::shared_ptr<ScanOptions>> options(plan.paths.size());
-
+    const WritePlan& plan, std::shared_ptr<ScanOptions> scan_options,
```

Comment:
```suggestion
    const WritePlan& plan,
```

Instead, `scan_options` should be derived from `plan.{schema,partitioning}` to project out columns implicit in the partitioning

Reply:
That's a task for [ARROW-8382](https://jira.apache.org/jira/projects/ARROW/issues/ARROW-8382). The writer shouldn't be aware of such issue, it takes a file source, a format, a RecordBatchReader (the dropping should happen internally here) and writes it­.

Since the Writer API has no bindings and not in used, I added this to ARROW-8382 so we can prioritise  ARROW-8062 (parquet dataset) and ARROW-8318 (fragment stored in dataset).